### PR TITLE
only allow pushes from master or whitelisted branches

### PR DESCRIPTION
**Pull request by gforsyth**
If I'm right, then we should know that this is the fix since it won't push to the `gh-pages` branch

### CODE REVIEW

**Comment by gforsyth** [comment]
Ok, this worked and I just tested on one of my repos, too. Once the merge commit clears CI I'll merge this and then do a new release

**Comment by gforsyth** [comment]
Ok, fool me once, etc...
I broke out the `canpush` logic into its own function and added some tests to make sure this doesn't happen again.  Also, as per @Carreau 's suggestion, changed `if branch_whitelist is None` to `if not branch_whitelist` in order to catch possibility that the whitelist is an empty set or list

File: doctr/travis.py [hunk 1]
```diff
@@ -146,32 +146,20 @@ def setup_GitHub_push(deploy_repo, auth_type='deploy_key', full_key_path='github
     For ``auth_type='deploy_key'``, this sets up the remote with ssh access.
     """
 
-    if branch_whitelist is None:
+    if not branch_whitelist:
         branch_whitelist={'master'}
 
     if require_master is not None:
```
Comment (Carreau) [inline doctr/travis.py]: do you want to actually remove the `require_master ` param here, it is not used anymore. 
If you are not comiting to Python api that would be a good thing.

File: doctr/travis.py [hunk 2]
```diff
@@ -401,3 +389,22 @@ def push_docs(deploy_branch='gh-pages'):
     run(['git', 'pull', 'doctr_remote', deploy_branch])
     print("Pushing commit")
     run(['git', 'push', '-q', 'doctr_remote', deploy_branch])
+
+def determine_push_rights(branch_whitelist):
+
+    canpush = True
+
+    TRAVIS_BRANCH = os.environ.get("TRAVIS_BRANCH", "")
+    TRAVIS_PULL_REQUEST = os.environ.get("TRAVIS_PULL_REQUEST", "")
```
Comment (Carreau) [inline doctr/travis.py]: I would have made `TRAVIS_BRANCH` and `TRAVIS_PULL_REQUEST ` parameters.

**Comment by gforsyth** [comment]
Oh. Yeah. In fact I think I have to do that or it'll break things

File: doctr/travis.py [hunk 3]
```diff
@@ -401,3 +389,22 @@ def push_docs(deploy_branch='gh-pages'):
     run(['git', 'pull', 'doctr_remote', deploy_branch])
     print("Pushing commit")
     run(['git', 'push', '-q', 'doctr_remote', deploy_branch])
+
+def determine_push_rights(branch_whitelist):
```
Comment (Carreau) [inline doctr/travis.py]: Should we test that `branch_whitelist` is a container (or at least not a string?)

File: doctr/travis.py (hunk 1 above)
Comment (gforsyth) [inline doctr/travis.py]: I don't think anyone is using the Python API -- let's remove it

Comment (gforsyth) [inline doctr/travis.py]: But maybe we can do that in a different PR?  I'd like to get this in before people who have `pip install doctr` start experiencing pain

File: doctr/travis.py (hunk 3 above)
Comment (gforsyth) [inline doctr/travis.py]: yes, but let's do that separately -- I think this is good enough for now, since the alternative is broken functionality

File: doctr/travis.py (hunk 1 above)
Comment (asmeurer) [inline doctr/travis.py]: I checked on GitHub and there are no `import doctr`s anywhere. I was thinking of using the Python API for a project, but I haven't implemented it yet. We shouldn't be too worried about changing it I'd say.

**Comment by Carreau** [comment]
> But maybe we can do that in a different PR? I'd like to get this in before people who have pip install doctr start experiencing pain

yes, as soon as travis is green, we can merge.

**Comment by gforsyth** [comment]
Starting the release now

**Comment by Carreau** [comment]
thanks.
